pkg/resources/server_test.go

```diff
@@ -0,0 +1,327 @@
+		It("server with plugin watcher enabled", func() {
+			fs := utils.FakeFilesystem{
+				Dirs:     []string{"sys/class/net/ib0/"},
+				Symlinks: map[string]string{"sys/class/net/ib0/device": "../../../0000:02:00.0"},
+			}
+			defer fs.Use()()
+			conf := &types.UserConfig{ResourceName: "test_server", RdmaHcaMax: 100, Devices: []string{"ib0"}}
+			obj, err := newResourceServer(conf, true, "rdma", "socket")
+			Expect(err).ToNot(HaveOccurred())
+			rs := obj.(*resourceServer)
+			Expect(rs.resourceName).To(Equal("rdma/test_server"))
+			Expect(rs.endPoint).To(Equal("test_server.socket"))
+			Expect(rs.watchMode).To(Equal(true))
+			Expect(len(rs.devs)).To(Equal(100))
+		})
+		It("server with plugin watcher disabled", func() {
+			fs := utils.FakeFilesystem{
+				Dirs:     []string{"sys/class/net/ib0/"},
+				Symlinks: map[string]string{"sys/class/net/ib0/device": "../../../0000:02:00.0"},
+			}
+			defer fs.Use()()
+			conf := &types.UserConfig{ResourceName: "test_server", RdmaHcaMax: 100, Devices: []string{"ib0"}}
+			obj, err := newResourceServer(conf, false, "rdma", "socket")
+			Expect(err).ToNot(HaveOccurred())
+			rs := obj.(*resourceServer)
+			Expect(rs.resourceName).To(Equal("rdma/test_server"))
+			Expect(rs.endPoint).To(Equal("test_server.socket"))
+			Expect(rs.watchMode).To(Equal(false))
+			Expect(len(rs.devs)).To(Equal(100))
+		})
+		It("server with plugin with invalid max number of resources", func() {
+			conf := &types.UserConfig{ResourceName: "test_server", RdmaHcaMax: -100, Devices: []string{"ib0"}}
+			obj, err := newResourceServer(conf, true, "rdma", "socket")
+			Expect(err).To(HaveOccurred())
+			Expect(obj).To(BeNil())
+		})
+	})
+	Context("Start", func() {
+		It("start server with plugin watcher enabled", func() {
+			grpcServer := grpc.NewServer([]grpc.ServerOption{}...)
+			rsc := &mocks.ResourceServerConnector{}
+			rsc.On("GetServer", mock.Anything).Return(grpcServer)
+			rsc.On("CreateServer").Return()
+			rsc.On("Listen", mock.Anything, mock.Anything).Return(nil, nil)
+			rsc.On("Serve", mock.Anything).Return()
+			rsc.On("Dial", mock.Anything, mock.Anything).Return(nil, nil)
+			rsc.On("Close", mock.Anything).Return()
+			rs := resourceServer{watchMode: true, rsConnector: rsc}
+			err := rs.Start()
+			Expect(err).ToNot(HaveOccurred())
+		})
+		It("start server with plugin watcher disabled", func() {
+			grpcServer := grpc.NewServer([]grpc.ServerOption{}...)
+			rsc := &mocks.ResourceServerConnector{}
+			rsc.On("GetServer", mock.Anything).Return(grpcServer)
+			rsc.On("CreateServer").Return()
+			rsc.On("Listen", mock.Anything, mock.Anything).Return(nil, nil)
+			rsc.On("Serve", mock.Anything).Return()
+			rsc.On("Dial", mock.Anything, mock.Anything).Return(nil, nil)
+			rsc.On("Close", mock.Anything).Return()
+			rsc.On("Register", mock.Anything, mock.Anything).Return(nil)
+			rs := resourceServer{watchMode: false, rsConnector: rsc}
+			err := rs.Start()
+			Expect(err).ToNot(HaveOccurred())
```

Comment:
can you ensure:
1. dial was called
2. Serve was called
3. resourceServer.register was called

Reply:
with rs.watchMode = false, all above must be called, even when try manually to eliminate the mock for any of them, the test faills.

Reply:
also if it passed the the error will be nil, and to pass all above must be called

Reply:
test need to be explicit not implicit

Reply:
for all mocked tests, there is the call of mockedObject.AssertExpectations(testCallsAssertionReporter) which will give error if the mocked function is not called in this and mentioned with On, 
e.g:
```
fake.On("GetServer").Return("something")
fake.AssertExpectations(testCallsAssertionReporter)
```
if fake.GetServer was not called an error will appear:
[root@r-cloudx3-07 k8s-rdma-shared-dev-plugin]# make test
running gofmt...
setting GOPATH...
running golangci-lint...
running tests...
Running Suite: Resources test suite
===================================
Random Seed: 1580279569
Will run 46 of 46 specs

••••••••2020/01/29 08:32:49 FAIL:       GetServer(string)
                at: [server_test.go:107 runner.go:113 runner.go:64 it_node.go:26 spec.go:215 spec.go:138 spec_runner.go:200 spec_runner.go:170 spec_runner.go:66 suite.go:62 ginkgo_dsl.go:226 ginkgo_dsl.go:207 resources_suite_test.go:29]
2020/01/29 08:32:49 FAIL: 0 out of 1 expectation(s) were met.
        The code you are testing needs to make 1 more call(s).
        at: [server_test.go:108 runner.go:113 runner.go:64 it_node.go:26 spec.go:215 spec.go:138 spec_runner.go:200 spec_runner.go:170 spec_runner.go:66 suite.go:62 ginkgo_dsl.go:226 ginkgo_dsl.go:207 resources_suite_test.go:29]
FAIL    github.com/Mellanox/k8s-rdma-shared-dev-plugin/pkg/resources    0.028s
FAIL
make: *** [test] Error 1